### ISSUE TITLE
No useless template expressions

### ISSUE THREAD

**Issue by MasterOfBob777** [eslint/eslint#14239]
**Please describe what the rule should do:**
This rule should detect useless template expressions (the `${ }`s inside a template string) like strings, numbers that don't require any calculations and other constant literals inside template expressions.

**What new ECMAScript feature does this rule relate to?**
Template Strings

**What category of rule is this? (place an "X" next to just one item)**

- [ ] Warns about a potential error (problem)
- [x] Suggests an alternate way of doing something (suggestion)
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Incorrect code
const foo = `${"bar"}`; // Warn
const bar = `${3}` // Warn, maybe an option for this to only include strings

// Correct
const baz = html`<div prop=${"thing"}> </div>` // Okay, because it has a tag function, maybe an option to add certain tag functions to a whitelist that will be checked
```

**Why should this rule be included in ESLint (instead of a plugin)?**
This rule fits the general style of eslint's base rules, and it goes well with other string related rules.

**Are you willing to submit a pull request to implement this rule?**
Not really, because of my general inexperience with the eslint codebase or eslint plugins, but I might find time to work on this

**Comment by ljharb** [comment]
I don't think it should be possible to warn on this for tagged templates, since the rule has no idea what the template is going to use it for (the author of the code calling it shouldn't need to know that either).

**Comment by MasterOfBob777** [comment]
I guess I should clarify that the possible option should be a whitelist of template tags, not a blacklist, because there aren't many actual uses for it but it could be useful in certain scenarios like a `md`` `tag or some other formats.

**Comment by ljharb** [comment]
Keying on the local name of the tag function - which could be literally anything, since a user can name it whatever they want - isn't going to be very reliable.

**Comment by mdjermanovic** [comment]
Hi @MasterOfBob777, thanks for the issue!

Unfortunately, we won't be able to accept this proposal. Per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies), we will only accept new rules when they relate to new ECMAScript features that reached stage 4 in the preceding 12 months, which isn't the case with template literals.

You can configure the [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule with the following selector to report template expressions that are literals:

```
":not(TaggedTemplateExpression) > TemplateLiteral > Literal.expressions"
```

Example:

```js
/* eslint no-restricted-syntax: ["error", {
    selector: ":not(TaggedTemplateExpression) > TemplateLiteral > Literal.expressions",
    message: "Unneeded template expression."
}]*/

const foo = `${"bar"}`; // error
const bar = `${3}`; // error

const baz = html`<div prop=${"thing"}> </div>`; // ok
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgc2VsZWN0b3I6IFwiOm5vdChUYWdnZWRUZW1wbGF0ZUV4cHJlc3Npb24pID4gVGVtcGxhdGVMaXRlcmFsID4gTGl0ZXJhbC5leHByZXNzaW9uc1wiLFxuICAgIG1lc3NhZ2U6IFwiVW5uZWVkZWQgdGVtcGxhdGUgZXhwcmVzc2lvbi5cIlxufV0qL1xuXG5jb25zdCBmb28gPSBgJHtcImJhclwifWA7IC8vIGVycm9yXG5jb25zdCBiYXIgPSBgJHszfWA7IC8vIGVycm9yXG5cbmNvbnN0IGJheiA9IGh0bWxgPGRpdiBwcm9wPSR7XCJ0aGluZ1wifT4gPC9kaXY+YDsgLy8gb2siLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

I'm closing this issue per our rules policies.
